### DISCUSSION TITLE
[BrowserKit] Managing multiple CookieJar on the same HttpClient

### DISCUSSION THREAD

**Discussion by paullallier** [symfony/symfony#51660] [Q&A]
### Description

I'd like to be able to test with BrowserKit HttpClient, simulating multiple users, but this means I need to swap the cookie jars around - at least if I want to avoid spawning lots of clients.

Would you consider allowing a setCookieJar() method or would this fundamentally break something?  Perhaps on AbstractBrowser?  It might even be easy enough for me to PR it!

**Comment by stof** [comment]
If you want to simulate multiple users, the right way is to instantiate multiple browsers (that's also how you would do it with actual browsers btw)

**Comment by paullallier** [comment]
That does seem more correct, but I'm struggling to see how I'd spawn extra clients in my test runner without breaking all the self::assert... calls, which look for the client with self::getClient().  But perhaps that's a discussion for Slack rather than here?

**Reply by stof** [reply]
I think we should add an optional AbstractBrowser argument on all those assertion methods to be able to use another one than the default one, relying on `self::getClient()` only when a browser is not provided.
Do you want to do such a contribution for 6.4 ?

**Reply by stof** [reply]
Note that in existing versions, you can already reuse the assertion constraints from the `Symfony\Component\BrowserKit\Test\Constraint` and `Symfony\Component\HttpFoundation\Test\Constraint` namespaces, using them with PHPUnit's `assertThat` yourselves. However, you'll indeed loose the special failure display for response assertions.

**Reply by paullallier** [reply]
Thanks stof.

In 6.3, I suspect I'll just patch Symfony (and Api-platform) so that I can hack the CookieJar - since it seems to need only a few small changes - but I think in the longer term having the assertions take an optional alternative AbstractBrowser would be useful (and let me drop my hack).

I'm certainly happy to have a go at implementing it - since it would be my first Symfony PR, it might prove beyond my skill level, but we'll see.  If I make a start and create a draft PR, would you be willing to sanity check it for me please?

**Reply by stof** [reply]
The core team will check PRs anyway (either me or someone else).

And if you need help to contribute, don't hesitate to join the `#contribs` channel on the Symfony Slack and ask for help there. The core team (and probably also some other contributors) will be there to answer.
